File: backend/app/models/vehicle_log.py
from sqlalchemy import Column, Integer, ForeignKey, Numeric, Boolean, Text, DateTime, PrimaryKeyConstraint
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from app.database import Base


class VehicleLog(Base):
    __tablename__ = "vehicle_logs"

    id = Column(Integer, autoincrement=True)
    vehicle_id = Column(
        Integer, ForeignKey("vehicles.id", ondelete="CASCADE"), nullable=False
    )

    # Battery & Power
    battery_voltage = Column(Numeric(5, 2))
    battery_current = Column(Numeric(6, 2))

    # Communication
    rssi = Column(Integer)

    # Flight Status
    mode = Column(Text)  # Flight mode (GUIDED, AUTO, MANUAL, etc)
    armed = Column(Boolean, default=False)
    guided = Column(Boolean, default=False)
    system_status = Column(Text)

    # Position (GPS)
    latitude = Column(Numeric(10, 8))
    longitude = Column(Numeric(11, 8))
    altitude = Column(Numeric(8, 2))  # Altitude in meters
    gps_fix = Column(Integer)  # GPS fix type (0=no fix, 1=2D, 2=3D, etc)

    # Navigation
    heading = Column(Numeric(5, 2))  # Compass heading (0-360 degrees)
    speed = Column(Numeric(6, 2))  # Speed in m/s

    # Orientation (IMU)
    roll = Column(Numeric(6, 2))  # Roll angle in degrees
    pitch = Column(Numeric(6, 2))  # Pitch angle in degrees
    yaw = Column(Numeric(6, 2))  # Yaw angle in degrees

    # Sensors
    temperature = Column(Numeric(5, 2))  # Temperature in Celsius

    created_at = Column(DateTime(timezone=True), server_default=func.now())

    __table_args__ = (
        PrimaryKeyConstraint("id", "created_at"),
    )

    vehicle = relationship("Vehicle", back_populates="vehicle_logs")

File: backend/internal/model/vehicle_log.go
package model

import "time"

// VehicleLog stores telemetry and status logs from vehicles
type VehicleLog struct {
	ID                uint      `json:"id" gorm:"primaryKey"`
	VehicleID         uint      `json:"vehicle_id" gorm:"not null;index"`
	Vehicle           *Vehicle  `json:"vehicle,omitempty" gorm:"foreignKey:VehicleID"`
	BatteryVoltage    *float64  `json:"battery_voltage" gorm:"type:numeric"`
	BatteryCurrent    *float64  `json:"battery_current" gorm:"type:numeric"`
	RSSI              *int      `json:"rssi" gorm:"type:int"`
	Mode              *string   `json:"mode" gorm:"type:varchar(50)"`
	Latitude          *float64  `json:"latitude" gorm:"type:numeric"`
	Longitude         *float64  `json:"longitude" gorm:"type:numeric"`
	Altitude          *float64  `json:"altitude" gorm:"type:numeric"`
	Heading           *float64  `json:"heading" gorm:"type:numeric"`
	Armed             *bool     `json:"armed" gorm:"type:boolean"`
	GPSok             *bool     `json:"gps_ok" gorm:"type:boolean"`
	SystemStatus      *string   `json:"system_status" gorm:"type:varchar(50)"`
	Speed             *float64  `json:"speed" gorm:"type:numeric"`
	Roll              *float64  `json:"roll" gorm:"type:numeric"`
	Pitch             *float64  `json:"pitch" gorm:"type:numeric"`
	Yaw               *float64  `json:"yaw" gorm:"type:numeric"`
	TemperatureSystem *string   `json:"temperature_system" gorm:"type:varchar(50)"` // Type field from schema
	CreatedAt         time.Time `json:"created_at" gorm:"autoCreateTime;index"`
}

func (VehicleLog) TableName() string {
	return "vehicle_logs"
}

// VehicleLogQuery for filtering logs
type VehicleLogQuery struct {
	VehicleID uint      `query:"vehicle_id"`
	StartTime time.Time `query:"start_time"`
	EndTime   time.Time `query:"end_time"`
	Limit     int       `query:"limit"`
	Offset    int       `query:"offset"`
}

// Request/Response Models for VehicleLog
type CreateVehicleLogRequest struct {
	VehicleID         uint     `json:"vehicle_id" example:"1"`
	BatteryVoltage    *float64 `json:"battery_voltage,omitempty" example:"12.5"`
	BatteryCurrent    *float64 `json:"battery_current,omitempty" example:"2.3"`
	RSSI              *int     `json:"rssi,omitempty" example:"-65"`
	Mode              *string  `json:"mode,omitempty" example:"AUTO"`
	Latitude          *float64 `json:"latitude,omitempty" example:"-6.2088"`
	Longitude         *float64 `json:"longitude,omitempty" example:"106.8456"`
	Altitude          *float64 `json:"altitude,omitempty" example:"10.5"`
	Heading           *float64 `json:"heading,omitempty" example:"90.5"`
	Armed             *bool    `json:"armed,omitempty" example:"true"`
	GPSok             *bool    `json:"gps_ok,omitempty" example:"true"`
	SystemStatus      *string  `json:"system_status,omitempty" example:"OK"`
	Speed             *float64 `json:"speed,omitempty" example:"5.2"`
	Roll              *float64 `json:"roll,omitempty" example:"0.5"`
	Pitch             *float64 `json:"pitch,omitempty" example:"1.2"`
	Yaw               *float64 `json:"yaw,omitempty" example:"90.5"`
	TemperatureSystem *string  `json:"temperature_system,omitempty" example:"Normal"`
}

